### ISSUE TITLE
FR: CoqIDE: moving cursor to beginning and end of file

### ISSUE THREAD

**Issue by eluuk** [coq/coq#9899]
#### Description of the problem

Are there shortcuts to put cursor to the beginning and end of file in CoqIDE w/out restarting it and compiling to file's end (respectively)? Usually I just want to look at the code elsewhere w/out changing the compilation state (and the processing causes a delay). Currently I'm using `xdotool` for this, which is almost as annoying (as it causes a short delay _and_ a fast scroll by sending multiple <kbd>PgUp</kbd> or <kbd>PgDown</kbd>)

#### Coq Version

8.9.0

**Comment by ppedrot** [comment]
Control + Home / End works for me.

**Comment by maximedenes** [comment]
@ppedrot that's because you have an ancient CoqIDE installation (that you updated but which kept the old defaults), or you customized the key bindings. Since around 8.6, the default binding for restarting Coq / processing the whole fifle are precisely CTRL + Home and CTRL + End, which is quite disastrous from a usability standpoint.

**Comment by eluuk** [comment]
Perhaps [`<Control>`](https://github.com/coq/coq/blob/1d68c24736b4cf68ac0c2f70122e3f3d28f0e876/ide/preferences.ml#L338) should be changed to `<Alt>`?

**Comment by ppedrot** [comment]
No, otherwise it messes with other shortcuts. I personally use CTRL+ALT for Coq document navigation modifiers, which doesn't interfere with the KDE shortcuts. I never quite understood why we settled for CTRL as the default modifier, but this exact topic seems to be a never ending debate.

**Comment by maximedenes** [comment]
> I personally use CTRL+ALT for Coq document navigation modifiers, which doesn't interfere with the KDE shortcuts

Yeah but it interferes with the shortcuts used by GNOME to switch virtual screens (or whatever they are called) by default...

> I never quite understood why we settled for CTRL as the default modifier, but this exact topic seems to be a never ending debate.

I don't think that's a matter of debate, we all agree we'd like bindings that don't clash on the main environments. I think the problem is more that we are incompetent in writing IDEs.

**Comment by eluuk** [comment]
> Perhaps [`<Control>`](https://github.com/coq/coq/blob/1d68c24736b4cf68ac0c2f70122e3f3d28f0e876/ide/preferences.ml#L338) should be changed to `<Alt>`?

Yep, it would be probably better left as it is, and have something like <kbd>Alt</kbd>+<kbd>Home</kbd> and <kbd>Alt</kbd>+<kbd>End</kbd> added

**Comment by eponier** [comment]
I switched to <kbd>Alt</kbd> a while ago and I have no problem with it. @ppedrot What shortcuts do you think of?

**Comment by herbelin** [comment]
For information, I made a proposal for going to the end without evaluating at #10006.

**Comment by eluuk** [comment]
I may have an idea on how to solve the problem. As it's known, a simple solution would be to change the default modifier of Navigation to e.g. <kbd>Alt</kbd>. However, this would affect other Navigation shortcuts, which many of us are accustomed to. But if `~/.config/coqide.keys` would override `~/.config/coqiderc`, changing only the bindings for Restart and Compile-to-End in the former would suffice. At the moment this works if you, additionally, comment out

`# modifier_for_navigation = "<Control>"`

in `~/.config/coqiderc` and make both files read-only. I find this hack quite convenient, but a permanent solution by way of `~/.config/coqide.keys` having priority over `~/.config/coqiderc` would solve the issue once and for all, with a complete control over CoqIDE's keyboard shortcuts as a bonus

**Comment by herbelin** [comment]
> I find this hack quite convenient, but a permanent solution by way of ~/.config/coqide.keys having priority over ~/.config/coqiderc would solve the issue once and for all, with a complete control over CoqIDE's keyboard shortcuts as a bonus

That `coqiderc` overrides changes from `coqide.keys` is a bug. I made #10014 for that.

The question remains whether we give a different default than <kbd>Ctrl</kbd><kbd>Home</kbd> for restart-coq, so that the pre-existing <kbd>Ctrl</kbd><kbd>Home</kbd> for go-to-start-of-buffer is not overriden (and similarly for <kbd>Ctrl</kbd><kbd>End</kbd>). If yes, can this alternative defaut be e.g. <kbd>Ctrl</kbd><kbd>Alt</kbd><kbd>Home</kbd>?


Side question, if anyone knows: where is it defined what the default effects of <kbd>Home</kbd> (go to start of line), <kbd>Ctrl</kbd><kbd>Home</kbd> (go to start of buffer), <kbd>End</kbd> (go to end of line), <kbd>Ctrl</kbd><kbd>End</kbd> (go to end of buffer) are? Is it a gtksourceview convention?

Other side question: would it make sense to have default settings dependent not only on the operating system but on the window manager? How is it done in other editors?

**Comment by eluuk** [comment]
On my system (Linux, Xfce4) neither <kbd>Alt</kbd><kbd>Home</kbd> nor <kbd>Ctrl</kbd><kbd>Alt</kbd><kbd>Home</kbd> do nothing, and similarly for <kbd>End</kbd>. I'm quite happy w/

**Comment by herbelin** [comment]
So, if you are happy with a custom solution (+ #10014), there is maybe no more need for a change?

I'm asking because I wonder whether it is worth to propose bindings for go-to-start and go-to-end in the μPG mode (#10006). This would then typically be Meta-Right and Meta-Left.

**Comment by eluuk** [comment]
Indeed, @herbelin, I wouldn't use <kbd>Alt</kbd><kbd>&rarr;</kbd>, etc. anyway. The only thing probably worth doing is to change the default bindings for Restart and Evaluate to End in `~/.config/coq/coqide.keys` and/or say something in the documentation (either about the change or the masking problem -- not sure if the problem is unique to Linux, but so it seems). As for the new bindings, I'm keeping my <kbd>Alt</kbd><kbd>Home</kbd> and <kbd>Alt</kbd><kbd>End</kbd>, but you can put there whatever else you want instead

**Comment by herbelin** [comment]
OK, let's hear for other opinions about whether a change of the default bindings for Restart and Evaluate to End is worth to be done developer-side or if only documenting how to change it user-side is enough.

In the meantime, I'm going to provide bindings for go-to-start and go-to-end in the emacs mode.

**Comment by eponier** [comment]
Personally, I also find <kbd>Ctrl</kbd><kbd>&#8594;</kbd> and <kbd>Ctrl</kbd><kbd>&#8592;</kbd> annoying for the exact same reason that they clash with other bindings. Indeed, I use them to move the cursor word by word. I think whatever the default is for the navigation keys, it will annoy some of the users. Using <kbd>Ctrl</kbd> everywhere is not ideal, but at least it is uniform. Thus, I would favour the simplest solution, i.e. just properly documenting how to change the default.
